pkg/controllers/readiness/wellknown_ready_controller.go

```diff
@@ -0,0 +1,351 @@
+	}
+
+	// pass the KAS service name for SNI
+	rt, err := transport.TransportFor("kubernetes.default.svc", caData, nil, nil)
+	if err != nil {
+		return false, "", fmt.Errorf("failed to build transport for SA ca.crt: %v", err)
+	}
+
+	ips, err := c.getAPIServerIPs()
+	if err != nil {
+		return false, "", fmt.Errorf("failed to get API server IPs: %v", err)
+	}
+
+	for _, ip := range ips {
+		wellknownReady, wellknownMsg, err := c.checkWellknownEndpointReady(ip, rt, route)
+		if err != nil || !wellknownReady {
+			return wellknownReady, wellknownMsg, err
+		}
+	}
+
+	return true, "", nil
+}
+
+func (c *wellKnownReadyController) checkWellknownEndpointReady(apiIP string, rt http.RoundTripper, route *routev1.Route) (bool, string, error) {
+	wellKnown := "https://" + apiIP + "/.well-known/oauth-authorization-server"
+
+	req, err := http.NewRequest(http.MethodGet, wellKnown, nil)
+	if err != nil {
+		return false, "", fmt.Errorf("failed to build request to well-known %s: %v", wellKnown, err)
+	}
+
+	resp, err := rt.RoundTrip(req)
+	if err != nil {
+		return false, "", fmt.Errorf("failed to GET well-known %s: %v", wellKnown, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return false, fmt.Sprintf("got '%s' status while trying to GET the OAuth well-known %s endpoint data", resp.Status, wellKnown), nil
+	}
+
+	var receivedValues map[string]interface{}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, "", fmt.Errorf("failed to read well-known %s body: %v", wellKnown, err)
+	}
+	if err := json.Unmarshal(body, &receivedValues); err != nil {
+		return false, "", fmt.Errorf("failed to marshall well-known %s JSON: %v", wellKnown, err)
+	}
+
+	expectedMetadata := getMetadataStruct(route)
+	if !reflect.DeepEqual(expectedMetadata, receivedValues) {
+		return false, fmt.Sprintf("the value returned by the well-known %s endpoint does not match expectations", wellKnown), nil
+	}
+
+	return true, "", nil
+}
+
+func getOAuthMetadata(host string) string {
+	stubMetadata := `
+{
+  "issuer": "https://%s",
+  "authorization_endpoint": "https://%s/oauth/authorize",
+  "token_endpoint": "https://%s/oauth/token",
+  "scopes_supported": [
+    "user:check-access",
+    "user:full",
+    "user:info",
+    "user:list-projects",
+    "user:list-scoped-projects"
+  ],
+  "response_types_supported": [
+    "code",
+    "token"
+  ],
+  "grant_types_supported": [
+    "authorization_code",
+    "implicit"
+  ],
+  "code_challenge_methods_supported": [
+    "plain",
+    "S256"
+  ]
+}
+`
+	return strings.TrimSpace(fmt.Sprintf(stubMetadata, host, host, host))
+}
+
+func getMetadataStruct(route *routev1.Route) map[string]interface{} {
+	var ret map[string]interface{}
+
+	metadataJSON := getOAuthMetadata(route.Spec.Host)
+	err := json.Unmarshal([]byte(metadataJSON), &ret)
+	if err != nil {
+		// should never happen unless the static metadata is broken
+		panic(err)
+	}
+
+	return ret
+}
```

Comment:
you could get rid of this duplication by listing the CM and comparing its content to what KAS returns. I can do that in a follow-up

Reply:
+1 for follow up